src/transformStateWithClones.js

```diff
@@ -7,7 +7,25 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  actions.forEach(action => {
+    if (action.type === 'clear') {
+      currentState = {};
+    } else if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
+      action.keysToRemove.forEach(key => {
+        delete currentState[key];
+      });
+    }
+    stateHistory.push(currentState);
```

Comment:
Instead of pushing the currentState into the stateHistory inside the forEach loop, which is repeated in every condition, perform this action once after the switch or if-else block to adhere to the DRY (Don't Repeat Yourself) principle.